### PR TITLE
GCP: set project id, use isolated config

### PR DESCRIPTION
**Pull request by Tener**
This PR fixes two issues with GCP integration:
- If a user has an existing `gcloud` installation, the configuration (in particular authorized accounts) can conflict with the configuration we set. Following the example set by Azure integration, we will use an isolated config under `~/.tsh/gcp`. This is done with `CLOUDSDK_CONFIG` env var.
- Without preexisting configuration, `tsh gcloud` will be missing the project id. There are multiple ways we can set that, but here we use the `CLOUDSDK_CORE_PROJECT` env variable.

### CODE REVIEW

File: tool/tsh/gcp_test.go [hunk 1]
```diff
@@ -258,3 +259,71 @@ func TestSortedGCPServiceAccounts(t *testing.T) {
 		})
 	}
 }
+
+func Test_gcpApp_Config(t *testing.T) {
```
Comment (smallinsky) [inline tool/tsh/gcp_test.go:263]:
```suggestion
func Test_GCPApp_Config(t *testing.T) {
```

File: tool/tsh/gcp.go [hunk 2]
```diff
@@ -126,9 +129,35 @@ func (a *gcpApp) Close() error {
...
+		return "", trace.BadParameter("cannot extract project ID from service account %q", serviceAccount)
+	}
+
+	for i, name := range projectIDRegexp.SubexpNames() {
+		if name == "projectID" {
+			return match[i], nil
+		}
+	}
+
+	return "", trace.BadParameter("cannot extract project ID from service account %q", serviceAccount)
```
Comment (smallinsky) [inline tool/tsh/gcp.go:150]: Any regexp instead of  string split ? Personally for me split is more readable that reading regxp expression

**Review by smallinsky — APPROVED** [review]

File: tool/tsh/gcp_test.go (hunk 1 above)
Comment (Tener) [reply tool/tsh/gcp_test.go:263]: I think the canonical name of the test should be `Test_{typeName}_{functionName}`? The type name is `gcpApp` and not `GCPApp`, as the type is not exported.

File: tool/tsh/gcp.go (hunk 2 above)
Comment (Tener) [reply tool/tsh/gcp.go:150]: Yeah, I was on the fence about this. I dislike regexps but this looked like a good use case. I can easily reimplement this using split, though.

**Review by greedy52 — APPROVED** [review]
Any other places that we can also do the validation of the service account name? e.g. when adding traits by tctl (seems we do validate azure identity there). Catching a bad service account name at `GetEnvVar` is pretty late in my opinion.

File: tool/tsh/gcp_test.go (hunk 1 above)
Comment (smallinsky) [reply tool/tsh/gcp_test.go:263]: Yea, I thought that I canceled this comment :-) I'm making it as resolved.

File: tool/tsh/gcp.go (hunk 2 above)
Comment (smallinsky) [reply tool/tsh/gcp.go:150]: Current version LGTM so this is only my personal bypass agains regex so leaving it up to you.

**Comment by Tener** [comment]
> Any other places that we can also do the validation of the service account name? e.g. when adding traits by tctl (seems we do validate azure identity there). Catching a bad service account name at `GetEnvVar` is pretty late in my opinion.

Agreed, we could add this to `tclt` and `tsh app login` both. For SSO-based accounts, `tsh app login` is probably the first user-facing step we can report a bad account such as this. I'll raise a separate PR for this.

File: tool/tsh/gcp.go (hunk 2 above)
Comment (Tener) [reply tool/tsh/gcp.go]: I gave it a go, and I like the non-regexp version better. It is easier to reason about and leads to much nicer error messages, which is important for validation.

File: tool/tsh/gcp.go [hunk 3]
```diff
@@ -126,9 +128,47 @@ func (a *gcpApp) Close() error {
...
+	if !found {
+		return "", trace.BadParameter("invalid service account format: missing <project-id>.iam.gserviceaccount.com after @")
+	}
+
+	if projectID == "" {
+		return "", trace.BadParameter("invalid service account format: missing project ID")
+	}
+
+	if iamDomain != "iam.gserviceaccount.com" {
+		return "", trace.BadParameter("invalid service account format: expected suffix %q, got %q", "iam.gserviceaccount.com", iamDomain)
```
Comment (GavinFrazar) [inline tool/tsh/gcp.go:158]: nit: move "iam.gserviceaccount.com" into a constant somewhere? Only because we reference that exact string 3 times

**Review by GavinFrazar — APPROVED** [review]

File: tool/tsh/gcp.go (hunk 3 above)
Comment (Tener) [reply tool/tsh/gcp.go:158]: I'm doing some refactoring for the upcoming PR dealing with additional validation of GCP service accounts in `tctl` and `tsh app login`, I'll move it to constant there.

Comment (Tener) [reply tool/tsh/gcp.go:158]: Done in https://github.com/gravitational/teleport/pull/20180.

**Pull request merged**